index.js

```diff
@@ -88,19 +90,59 @@ export default class Carousel extends Component {
     if (this.state.childrenLength) {
       this._setUpTimer();
     }
+    // Set up pages but not content. Content will be set up via onLayout event.
+    this._setUpPages();
   }
 
   componentWillUnmount() {
     this._clearTimer();
   }
 
   componentWillReceiveProps(nextProps) {
-    let childrenLength = 0;
-    if (nextProps.children) {
-      childrenLength = nextProps.children.length ? nextProps.children.length : 1;
+    if (this.props.children !== nextProps.children) {
+        let childrenLength = 0;
+        if (nextProps.children) {
+          const length = React.Children.toArray(nextProps.children).length;
+          childrenLength = length || 1;
+        }
+        this.setState({ childrenLength, contents: null }, () =>
```

Comment:
Allow initial render to complete fast and not lock up UI thread. When carousel contents are ready, we'll render again.